sound/soc/sof/intel/hda.c

```diff
@@ -297,6 +307,268 @@ static const struct sof_intel_dsp_desc *get_chip_info(int pci_id)
+	sdev->bar[HDA_DSP_DRSM_BAR] = bus->drsmcap;
+
+	return 0;
+}
+
+#define IDISP_INTEL_VENDOR_ID	0x80860000
+
+/*
+ * load the legacy codec driver
+ */
+static void sof_load_codec_module(struct hda_codec *codec)
+{
+#ifdef MODULE
+	char modalias[MODULE_NAME_LEN];
+	const char *mod = NULL;
+
+	snd_hdac_codec_modalias(&codec->core, modalias, sizeof(modalias));
+	mod = modalias;
+	dev_dbg(&codec->core.dev, "loading %s codec module\n", mod);
+	request_module(mod);
+#endif
+}
+
+/*
+ * Probe the given codec address
+ */
+static int sof_probe_hda_codec(struct hda_bus *hbus, int addr)
+{
+	unsigned int cmd = (addr << 28) | (AC_NODE_ROOT << 20) |
+		(AC_VERB_PARAMETERS << 8) | AC_PAR_VENDOR_ID;
+	unsigned int res = -1;
+	struct hdac_hda_priv *hda_codec;
+	struct hdac_device *hdev;
+	int err;
+
+	mutex_lock(&hbus->core.cmd_mutex);
+	snd_hdac_bus_send_cmd(&hbus->core, cmd);
+	snd_hdac_bus_get_response(&hbus->core, addr, &res);
+	mutex_unlock(&hbus->core.cmd_mutex);
+	if (res == -1)
+		return -EIO;
+	dev_dbg(hbus->core.dev, "codec #%d probed OK: %x\n", addr, res);
+
+	hda_codec = devm_kzalloc(&hbus->pci->dev, sizeof(*hda_codec),
+				 GFP_KERNEL);
+	if (!hda_codec)
+		return -ENOMEM;
+
+	hda_codec->codec.bus = hbus;
+	hdev = &hda_codec->codec.core;
+
+	err = snd_hdac_ext_bus_device_init(&hbus->core, addr, hdev);
+	if (err < 0)
+		return err;
+
+	/* use legacy bus only for HDA codecs, idisp uses ext bus */
+	if ((res & 0xFFFF0000) != IDISP_INTEL_VENDOR_ID) {
+		hdev->type = HDA_DEV_LEGACY;
+		sof_load_codec_module(&hda_codec->codec);
+	}
+	return 0;
+}
+
+/* Codec initialization */
+static void sof_hda_codec_create(struct snd_sof_dev *sdev)
+{
+	struct hda_bus *hbus = sdev->hbus;
+	struct hdac_bus *bus = &hbus->core;
+	int c, max_slots;
+
+	max_slots = HDA_MAX_CODECS;
+
+	/* First try to probe all given codec slots */
+	for (c = 0; c < max_slots; c++) {
+		if ((bus->codec_mask & (1 << c))) {
+			if (sof_probe_hda_codec(hbus, c) < 0) {
+				/*
+				 * Some BIOSen give you wrong codec addresses
+				 * that don't exist
+				 */
+				dev_warn(bus->dev,
+					 "Codec #%d probe error; disabling it...\n", c);
+				bus->codec_mask &= ~(1 << c);
+				/*
+				 * More badly, accessing to a non-existing
+				 * codec often screws up the controller bus,
+				 * and disturbs the further communications.
+				 * Thus if an error occurs during probing,
+				 * better to reset the controller bus to get
+				 * back to the sanity state.
+				 */
+				snd_hdac_bus_stop_chip(bus);
+				sof_hda_init_chip(sdev, true);
+			}
+		}
+	}
+}
+
+static int sof_i915_init(struct hdac_bus *bus)
+{
+	int err;
+
+	/*
+	 * The HDMI codec is in GPU so we need to ensure that it is powered
+	 * up and ready for probe
+	 */
+	err = snd_hdac_i915_init(bus);
+	if (err < 0)
+		return err;
+
+	err = snd_hdac_display_power(bus, true);
+	if (err < 0)
+		dev_err(bus->dev, "Cannot turn on display power on i915\n");
+
+	return err;
+}
+
+#else
+static int sof_first_init(struct snd_sof_dev *sdev)
```

Comment:
could we have an .ops here and select which mode we want? It'd allow all the HDA link part to go in a different file. Ifdefs are hard to handle at some point, it's difficult to read.

Reply:
OK, will change it, thanks.